connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java

```diff
@@ -1957,6 +1789,61 @@ private void verifyTaskHeaderConverter() {
         verify(plugins).newHeaderConverter(any(AbstractConfig.class), eq(WorkerConfig.HEADER_CONVERTER_CLASS_CONFIG), eq(ClassLoaderUsage.CURRENT_CLASSLOADER));
     }
 
+    private void mockGenericIsolation() {
+        when(plugins.connectorLoader(anyString())).thenReturn(pluginLoader);
+        when(plugins.withClassLoader(pluginLoader)).thenReturn(loaderSwap);
+    }
+
+    private void verifyGenericIsolation() {
+        verify(plugins, atLeastOnce()).withClassLoader(pluginLoader);
+        verify(loaderSwap, atLeastOnce()).close();
+    }
+
+    private void mockConnectorIsolation(String connectorClass, Connector connector) {
+        mockGenericIsolation();
+        when(plugins.newConnector(connectorClass)).thenReturn(connector);
+        when(connector.version()).thenReturn("1.0");
+    }
+
+    private void verifyConnectorIsolation(Connector connector) {
+        verifyGenericIsolation();
+        verify(plugins).newConnector(anyString());
+        verify(connector, atLeastOnce()).version();
+    }
+
+    private void mockTaskIsolation(Class<? extends Connector> connector, Class<? extends Task> taskClass, Task task) {
+        mockGenericIsolation();
+        doReturn(connector).when(plugins).connectorClass(connector.getName());
+        when(plugins.newTask(taskClass)).thenReturn(task);
+        when(task.version()).thenReturn("1.0");
+    }
+
+    private void verifyTaskIsolation(Task task) {
+        verifyGenericIsolation();
+        verify(plugins).connectorClass(anyString());
+        verify(plugins).newTask(any());
+        verify(task).version();
+    }
+
+    private void mockExecutorSubmit(boolean startRunnable) {
```

Comment:
I found this pretty confusing in the case when `startRunnable` is `true`, since we don't actually end up mocking anything on the executor itself, and the implicit expectation in that case is that the `Worker` was actually instantiated with a non-mocked executor.

Anything we can do to make this a bit less of a footgun? Javadoc, a different name, split out the two cases into different methods, mock the executor in either case, etc.?

Reply:
I split the methods out while keeping the sometimes-mocked-sometimes-not executorService. If using the real/fake executor service and the real/fake mock in sync is too tedious, we can follow this up with a cleanup that removes the real executor tests and always injects a managed executor service.

Reply:
Also it happened to coincide that the real-executorService was only used with startConnector, and the fake-executorService was only used with startTask. This appeared to be because startTask doesn't have a callback which depends on the execution of the WorkerTask, while startConnector does. If in the future the startTask does have a callback, then we will need to provide real executor functionality for the startTask tests.